core/src/main/scala/org/apache/spark/resource/ResourceUtils.scala

```diff
@@ -124,6 +124,35 @@ private[spark] object ResourceUtils extends Logging {
       .filter(_.amount > 0)
   }
 
+  // Used to take a fraction amount from a task resource requirement and split into a real
+  // integer amount and the number of parts expected. For instance, if the amount is 0.5,
+  // the we get (1, 2) back out.
+  // Returns tuple of (amount, numParts)
+  def calculateAmountAndPartsForFraction(amount: Double): (Int, Int) = {
+    val parts = if (amount <= 0.5) {
+      Math.floor(1.0 / amount).toInt
+    } else if (amount % 1 != 0) {
+      throw new SparkException(
+        s"The resource amount ${amount} must be either <= 0.5, or a whole number.")
+    } else {
+      1
```

Comment:
I am trying to understand how this method is used, and what it means.
If possible, can you point me to some info ? Thx
(I see that it is a refactoring of inline code to a method, but was wondering what it was actually trying to do).

Reply:
Sure it needs more documentation then, I'll update the comment to have more context and perhaps clarify the names a bit.  
This is used for task resources (doesn't apply to executor resources, only task level and only supported when value < 1, we don't allow 2.33 for instance) when you have fractional amounts.   For instance you can set spark.task.resource.gpu.amount=0.33 and what this really means is that I want 3 tasks to use the same GPU.
so do that inside Spark we convert this into an amount and a number of parts per resource.  Really number of parts you can think of as the number of slots per GPU address.   For 0.33 the amount = 1 and the numparts becomes 3. If the  spark.task.resource.gpu.amount=2 the amount =2 and numparts = 1.   So numparts is the number of ways to subdivide a resource address when the amount = 1.

Reply:
note it ends up being used stored here: https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/resource/ResourceUtils.scala#L64

It ultimately ends up being used here: https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/resource/ResourceAllocator.scala#L33 
where we track the allocations of each resource.  this allows for a single address to be used multiple times

Reply:
I attempted to clarify the comment here:

https://github.com/apache/spark/pull/27313/commits/d270a731629ca53fb14a32fb2f17e732cd9a16a1

Really I should probably go through everywhere that is used and try to clarify but I think that would be separate PR

Reply:
I am slightly confused about this.

If gpus on an executor is 2 and spark.task.resource.gpu.amount=0.33, is it invalid configuration 
 ? Or does it mean we run only 2 tasks on that node for that resource profile ?

If spark.task.resource.gpu.amount=1 and executor  has 2 gpu's, does it mean the task needs both gpu's ? Or only  1 ?

What does it mean to have spark.task.resource.gpu.amount > 1 ?

Reply:
> If gpus on an executor is 2 and spark.task.resource.gpu.amount=0.33, is it invalid configuration

Yes its valid, executor has 2 gpus and you allow 3 tasks on each gpu, means you can run 6 total tasks in parallel (assuming you have 6 cores and spark.task.cpus=1).

> If spark.task.resource.gpu.amount=1 and executor has 2 gpu's, does it mean the task needs both gpu's ? Or only 1 ?

It means each task gets 1 gpu, so if executor has 2 gpus you can run 2 tasks. 

if you have spark.task.resource.gpu.amount > 1 that means each task gets that amount of GPUs.   Meaning executor gpus=4 and spark.task.resource.gpu.amount=2, means each task gets 2 gpus and you have 2 tasks total on that executor.

If you configure spark.task.resource.gpu.amount=4 and spark.executor.resource.gpu.amount=1 - spark throws an error as its invalid configuration.

The idea was spark.task.resource.gpu.amount acts just like spark.task.cpus - with the addition it also supports the fractional amounts in case you want to run multiple tasks on a single gpu.